src/game/WorldSession.cpp

```diff
@@ -102,17 +99,16 @@ WorldSession::~WorldSession()
         LogoutPlayer(true);
 
     /// - If have unclosed socket, close it
-    if (m_Socket)
-    {
-        m_Socket->CloseSocket();
-        m_Socket->RemoveReference();
-        m_Socket = nullptr;
-    }
+    if (!m_Socket->IsClosed())
+        m_Socket->Close();
 
     ///- empty incoming packet queue
-    WorldPacket* packet = nullptr;
-    while (_recvQueue.next(packet))
+    while (!m_packetQueue.empty())
```

Comment:
Nice to see this being changed in something that is a bit better supported.

It is a lot of code so I am just going to ask.
Are you sure nothing is being written to the queue at this point in time?

You close the socket before you cleanup this queue but the actual callbacks are triggered by a network thread.
Are you sure the network thread is not going to make calls to this object during its deletion/after it has been deleted?

Looking at your socket code I am fairly certain this albeit with a very small frequency might still result in either a crash and/or memory leaking.

Btw if you are making the switch to boost anyway, you could also consider using its lockfree queue to get rid of the read/write lock for this queue.